question-servers/freeform.js

```diff
@@ -624,10 +640,62 @@ module.exports = {
             data.options.question_path = context.question_dir;
             data.options.client_files_question_path = path.join(context.question_dir, 'clientFilesQuestion');
 
-            module.exports.processQuestion('render', pc, data, context, (err, courseIssues, _data, html, _fileData, renderedElementNames) => {
-                if (ERR(err, callback)) return;
-                callback(null, courseIssues, html, renderedElementNames);
-            });
+            // This function will render the panel and then cache the results
+            // if cacheKey is not null
+            const doRender = (cacheKey) => {
+                module.exports.processQuestion('render', pc, data, context, (err, courseIssues, _data, html, _fileData, renderedElementNames) => {
+                    if (ERR(err, callback)) return;
+                    if (cacheKey) {
+                        cache.set(cacheKey, {
+                            courseIssues,
+                            html,
+                            renderedElementNames,
+                        });
+                    }
+                    callback(null, courseIssues, html, renderedElementNames, false /* cache miss */);
+                });
+            };
+
+            // This function will check the cache for the specified cache key
+            // and either return the cached render for a cache hit, or render
+            // the panel for a cache miss
+            const getFromCacheOrRender = (cacheKey) => {
+                cache.get(cacheKey, (err, cachedData) => {
+                    // We don't actually want to fail if the cache has an error; we'll
+                    // just render the panel as normal
+                    if (ERR(err, callback)) logger.error(err);
```

Comment:
I think this should be `if (ERR(err, () => {}) logger.error(err);` so that we won't call `callback`?